keps/sig-network/1864-disable-lb-node-ports/README.md

```diff
@@ -136,6 +146,182 @@ re-enabling node port should not cause any traffic disruptions.
 and not the `allocateLoadBalancerNodePorts` field.
 
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/20190731-production-readiness-review-process.md.
+
+The production readiness review questionnaire must be completed for features in
+v1.19 or later, but is non-blocking at this time. That is, approval is not
+required in order to be in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: ServiceLBNodePortControl
+    - Components depending on the feature gate: kube-apiserver
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+
+No, enabling the feature gate but not setting `spec.allocateLoadBalancerNodePorts` will not
+change any default behaviors in Service.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+
+Yes, if node ports are disabled, they can be re-enabled at any point by unsetting `spec.allocateLoadBalancerNodePorts`.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+This field will be alpha feature gated for 1 release before being enabled by default, this
+ensures the API can safely handle rollback if the field has been set already.
+
+* **Are there any tests for feature enablement/disablement?**
+
+Yes, there will be unit tests for the Service API strategy which exercises the behavior
+with the feautre gate enabled and disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+
+TODO before feature goes GA
```

Comment:
Given Beta is enabled by default, I actually claim this should be filled in for Beta.

Reply:
Same below.

Reply:
Noted, interestingly this feature doesn't have a "beta" stage. Since we are adding a new field to Service, the "alpha" stage is to ensure safe rollback by allowing apiserver to handle the new field 1 release before it is allowed by default. In this case should these questions be answered for alpha or for GA?

Reply:
Why there is no Beta?
I think this feature should go through all Alpha, Beta and GA phases. I don't see a good reason for not having Beta...

@johnbelamaric @deads2k - if you have thoughts on it

Reply:
> interestingly this feature doesn't have a "beta" stage

why would it not have a beta stage?

* alpha is disabled by default for a release, to allow early adopters to try it out and get n-1 API servers able to round-trip the data
* beta is typically enabled by default for at least one release, which exposes the feature to most clusters by allows conservative clusters to disable the feature until it has soaked in public
* GA enables unconditionally

Reply:
Seem I'm not alone in my opinion - as I wrote above: "I don't see a good reason for not having Beta..."

Reply:
I don't see any harm in having a beta stage and I have no objections to adding one. 

I had thought this was the precedent for new fields to v1 APIs (like Service) though since a field in a GA API that is on by default couldn't be removed or changed anyways? I guess a follow-up question I have is: 

Can a new field to a v1 API be deprecated during the beta stage? Or in this case is beta just to allow conservative users to opt-out like Jordan mentioned? In this case the feature itself is to opt-out of a default behavior (node ports for Service Type=LB) so I don't see this being likely.

Reply:
> Can a new field to a v1 API be deprecated during the beta stage?

That hasn't come up before... I wouldn't expect to introduce a beta field to a GA type unless there was a very clear path to GA agreed on.

> Or in this case is beta just to allow conservative users to opt-out like Jordan mentioned?

Opting-out or disabling if something goes wrong that isn't discovered until after the beta release.

Reply:
> I wouldn't expect to introduce a beta field to a GA type unless there was a very clear path to GA agreed on.

This was my impression as well which is why I had assumed the beta stage was redundant initially. I agree regardless that we should have a beta stage here for reasons you already mentioned. I'll update the KEP shortly.

Reply:
Updated graduation criteria to include beta.